aten/src/ATen/native/TensorShape.cpp

```diff
@@ -563,6 +563,82 @@ static void fastCatOutDim0(const Tensor& out, const MaterializedITensorListRef&
   TORCH_CHECK(outBytes == totalBytes);
 }
 
+static void parallelFastCatOutDim0(
+    const Tensor& out,
+    const MaterializedITensorListRef& inputs) {
+  // Partition inputs into parallel chunks based on number of elements
+  size_t out_elements = out.numel();
+  if (out_elements < at::internal::GRAIN_SIZE || at::get_num_threads() == 1) {
+    fastCatOutDim0(out, inputs);
+    return;
+  }
+  size_t total_elements = 0;
+  size_t copy_elements = 0;
+  std::vector<std::pair<size_t, size_t>> input_chunks;
+  size_t start = 0;
+  size_t end = 0;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    const Tensor& input = inputs.at(i);
+    const size_t input_elements = input.numel();
+    total_elements += input_elements;
+    TORCH_CHECK(out_elements >= total_elements);
+    if (i == inputs.size() - 1) {
+      TORCH_CHECK(out_elements == total_elements);
+    }
+    copy_elements += input_elements;
+    if (copy_elements >= at::internal::GRAIN_SIZE) {
+      end = i + 1;
+      input_chunks.emplace_back(start, end);
+      start = end;
+      copy_elements = 0;
+    }
+  }
+  if (start < inputs.size()) {
+    // The remaining elements < GRAIN_SIZE, but still have to be copied
+    input_chunks.emplace_back(start, inputs.size());
+  }
+
+  // Record starting bytes offset
+  size_t out_bytes = out.nbytes();
+  size_t total_bytes = 0;
+  std::vector<size_t> out_bytes_offset;
+  out_bytes_offset.emplace_back(0);
```

Comment:
please reserve when you know the size of the vector; in this case it's inputs.size() + 1